feat: utilise pytest for `test_kernel`

tc85324 opened the pull request:
### PR Type
Closes #542

- Feature
- Refactoring (no functional changes)
- Tests

### Description
Significantly reduces the LOCs and enhances modularity of the tests. This should help improve test consistency and reduce the implementation burden for testing any future additions to `kernel.py`.

### How Has This Been Tested?
The new tests pass.

### Does this PR introduce a breaking change?
Removes the following redundant/ineffective tests:
- `test_approximator_valid`: test did not perform effective checks
- `test_stein_kernel_computation`: redundant overlap with `test_stein_kernel_element_computation`
- `test_stein_kernel_invalid_base_kernel_missing_divergence` and `test_stein_kernel_invalid_base_kernel_no_pytree` are replaced with a single check of `InvalidKernel`

Modifies the following logic:
- ~~`SteinKernel` now only checks on a call to `compute_elementwise` if the `base_kernel` attribute is an instance of `Kernel` (as required)~~



### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

rg936672 commented:
This has merge conflicts atm

tc85324 commented:
Check that this covers #237

rg936672 commented on coreax/kernel.py:
```diff
@@ -915,6 +901,10 @@ def compute_elementwise(
         :param y: Vector :math:`\mathbf{y} \in \mathbb{R}^d`
         :return: Kernel evaluated at (``x``, ``y``)
         """
+        if not isinstance(self.base_kernel, Kernel):
```
As far as I'm aware, the preferred approach on Coreax at the moment is not to rely on `isinstance` checks, but rather to rely on `hasattr` checks (i.e. relying on duck typing rather than explicit typing). I'm happy for this change to go through, but you should probably get a second opinion!

tc85324 replied:
What do you think @tp832944 ?

tp832944 replied:
The agreed method (see https://github.com/gchq/coreax/issues/420#issuecomment-1961033259) is to just let it crash when safe to do so. Otherwise, use `try ... except`. If the error message from letting it crash is unhelpful, use `try ... except` then reraise a useful error. In this case, the attribute error raised by the next line (908) is fairly informative, so just delete 904-907.

rg936672 replied:
On the topic of helpful vs unhelpful error messages, I seem to remember that trying to JIT an object missing `tree_flatten` or `tree_unflatten` gave a really unhelpful error message. However, if we're moving to `equinox` this shouldn't be an issue.

tc85324 replied:
I'll remove the validation change from this pull and we can have any other discussions regarding validation in #583

rg936672 commented on tests/unit/test_kernel.py:
```diff
@@ -19,1993 +19,572 @@
-            length_scale=self.default_length_scale
-        )
-
-    def test_approximator_valid(self) -> None:
```
We seem to have lost this test!

tc85324 replied:
I can add this back in, but it will be deleted again if #583 is merged first as the functionality being tested will no longer exists (in this form). I had also been sceptical of the utility of this test, as it is only checking that the mocked approximator is being called, not that approximation is working (although I guess this is the only way to to have the test success decoupled from the success of the approximator implementation)

rg936672 reviewed (changes requested):
As well as the couple of comments I left, the main issue is that we've lost most of the test docstrings in the transition to pytest - if these are added back (as well as the missing test) this should be good to merge. This is nonetheless an impressive shortening!

tc85324 commented:
I've removed the validation, added back in the docstrings (where appropriate), and have enhanced the accuracy (and hopefully clarity) of the typehints.

rg936672 commented:
Oops. test errors on 3.9 and 3.10 (`TypeError: Multiple inheritance with NamedTuple is not supported`)

rg936672 reviewed (approved):
Once the test collection failures are fixed (should be easily resolved by making `_Problem` not inherit from `NamedTuple`), this looks good to go! 

In future, though, I'd ask that you don't amend or rebase commits that have already been reviewed if possible - it makes reviewing incremental changes (i.e. changes only since the last commit reviewed) more of a hassle.

tc85324 commented:
Sorry about that, I had realised it would be a problem straight after pushing. I've removed the generic from the `_Problem` tuple to fix collection on versions less than 3.11.
